src/joint_trajectory_action/joint_trajectory_action.py

```diff
@@ -488,3 +490,116 @@ def check_goal_state():
             self._result.error_code = self._result.GOAL_TOLERANCE_VIOLATED
             self._server.set_aborted(self._result)
         self._command_stop(goal.trajectory.joint_names, end_angles, start_time, dimensions_dict)
+
+    def _get_trajectory_parameters(self, joint_names):
+        # For each input trajectory, if path, goal, or goal_time tolerances
+        # provided, we will use these as opposed to reading from the
+        # parameter server/dynamic reconfigure
+
+        # Goal time tolerance - time buffer allowing goal constraints to be met
+        self._goal_time = self._dyn.config['goal_time']
+
+        # Stopped velocity tolerance - max velocity at end of execution
+        self._stopped_velocity = self._dyn.config['stopped_velocity_tolerance']
+
+        # Path execution and goal tolerances per joint
+        for jnt in joint_names:
+            if not jnt in self._limb.joint_names():
+                rospy.logerr(
+                    "%s: Trajectory Aborted - Provided Invalid Joint Name %s" %
+                    (self._action_name, jnt,))
+                self._result.error_code = self._result.INVALID_JOINTS
+                self._server.set_aborted(self._result)
+                return
+            # Path execution tolerance
+            path_error = self._dyn.config[jnt + '_trajectory']
+            self._path_thresh[jnt] = path_error
+
+            # Goal error tolerance
+            goal_error = self._dyn.config[jnt + '_goal']
+            self._goal_error[jnt] = goal_error
+
+            # PID gains if executing using the velocity (integral) controller
+            if self._mode == 'velocity':
+                self._pid[jnt].set_kp(self._dyn.config[jnt + '_kp'])
+                self._pid[jnt].set_ki(self._dyn.config[jnt + '_ki'])
+                self._pid[jnt].set_kd(self._dyn.config[jnt + '_kd'])
+                self._pid[jnt].initialize()
+
+    def _on_joint_trajectory(self, trajectory):
+        joint_names = trajectory.joint_names
```

Comment:
As this trajectory callback can execute at the same time as the Action callback, I think it would make sense to have a mutex lock on both to prevent simultaneous joint commands from being sent into the robot.